### PR TITLE
Concurrent progress bars

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

Implements concurrent progress bars. Resolves https://github.com/astral-sh/uv/issues/1209.

## Test Plan

https://github.com/astral-sh/uv/assets/34988408/b21bdfbb-8817-4873-a65c-16c9e8c7c460

### CODE REVIEW

**Comment by T-256** [comment]
could it be sorted (the largest at top)?

**Comment by ibraheemdev** [comment]
@T-256 I put the largest at the bottom which looks more natural, in my opinion.


https://github.com/astral-sh/uv/assets/34988408/52915706-5349-48b4-82e4-a72f8b1c9846

**Comment by charliermarsh** [comment]
The code generally looks good (well done) but need to find a bit of time to play with this myself to see if I have any feedback on the UX.

**Comment by ibraheemdev** [comment]
Just tested this on a large requirements file, I think any downloads <250kb shouldn't be displayed. The output changes too fast for it to be useful.

**Comment by ibraheemdev** [comment]
We also probably want a way to disable this output.

**Comment by charliermarsh** [comment]
We could consider making it opt-in. (That would also make it less risky to ship.)

**Comment by zanieb** [comment]
What's the risk, specifically? It seems improbable that this will break people's workflows right?

**Comment by zanieb** [comment]
We do have a preview option now though, if you really want to gate it. A dedicated setting might make sense too? but opt-in seems wrong.

**Comment by charliermarsh** [comment]
I think the risk is just that it’s overly verbose for most package installs and may require tuning. I would be fine shipping it under preview so we can get feedback on the UI/UX. I haven’t had a chance to play with it myself yet though.

**Comment by zanieb** [comment]
I don't mind iterating on display ux without opt-in, it feels much safer than other changes we make and we release very often.

**Comment by charliermarsh** [comment]
Yes, from me at least - not sure if others are planning to review.

**Comment by zanieb** [comment]
I'd love to get this merged, I don't think it's great to have it hanging for weeks. Anything I can do to move this forward?

**Comment by zanieb** [comment]
I did test it and was happy with it but I can do more investigation.

@ibraheemdev can you rebase the branch?

**Comment by charliermarsh** [comment]
I will review and merge today.

**Comment by ibraheemdev** [comment]
I'll rebase. I am a little concerned about the output when you have a lot of small downloads though, e.g. https://github.com/astral-sh/uv/issues/2706#issuecomment-2115781002. Even kitty was flickering a bit. We can probably buffer updates to the terminal a bit.

**Comment by zanieb** [comment]
If anyone has "good" test cases post 'em here so we have some common ground when considering the behavior.

**Comment by zanieb** [comment]
I'm also supportive of including an opt-out environment variable.

**Comment by charliermarsh** [comment]
I would still be fine making it opt-in if we want to get more feedback before stabilizing :)

**Comment by zanieb** [comment]
Testing with

```
cargo run -q -- pip install anyio --reinstall --no-cache
cargo run -q -- pip install pydantic --reinstall --no-cache
cargo run -q -- pip install prefect --reinstall --no-cache
cargo run -q -- pip install pyqt5 --reinstall --no-cache
```

All good to me

**Comment by T-256** [comment]
LGTM and as Charlie said, it could be opt-in via `--preview` and after stabilization make it default behavior. Also, as Zanie said we can have an option to opt-out from it.

> We can probably buffer updates to the terminal a bit.

Let's get some feedback from users, then we can set this interval value with more confidence.

**Comment by charliermarsh** [comment]
Should we just move behind preview to unblock the decision? @ibraheemdev do you _want_ it to ship to stable?

**Comment by ibraheemdev** [comment]
I'm fine with merging this behind preview or an opt-in flag, as long as it's accessible because it can be useful for debugging.

**Comment by zanieb** [comment]
I have a medium preference for shipping it in stable with an opt-out, but that requires a separate mechanism than `--preview` 🤷‍♀️

**Comment by charliermarsh** [comment]
Let's just move it to `--preview` to get it unblocked I think.

**Comment by charliermarsh** [comment]
I like how pip's progress bars are left-aligned and have a default width -- as-is the display is too wide IMO on my monitor:

Screenshot 2024-05-23 at 11 11 30 PM

Screenshot 2024-05-23 at 11 12 18 PM

How hard is it to achieve something like that?

**Comment by charliermarsh** [comment]
To achieve that, I guess we'd need to put the package name above the bar, or to the right of it, to achieve consistent alignment?

**Comment by charliermarsh** [comment]
Or use a fixed size for the package name. I can play with it.

**Comment by charliermarsh** [comment]
Here's a slightly different layout. The colors are washed out in the video but it's green on dim:

https://github.com/astral-sh/uv/assets/1309177/b7ada082-cb52-4553-9b1c-1d2e52d1bc62

**Comment by charliermarsh** [comment]
Thoughts?

File: crates/uv-distribution/src/distribution_database.rs [hunk 1]
```diff
@@ -449,7 +487,17 @@ impl<'a, Context: BuildContext> DistributionDatabase<'a, Context> {
                 // Download and unzip the wheel to a temporary directory.
                 let temp_dir = tempfile::tempdir_in(self.build_context.cache().root())
                     .map_err(Error::CacheWrite)?;
-                uv_extract::stream::unzip(&mut hasher, temp_dir.path()).await?;
+
+                match self.reporter {
+                    Some(ref reporter) => {
+                        let mut reader =
+                            ProgressReader::new(&mut hasher, progress.unwrap(), &**reporter);
```
Comment (charliermarsh) [inline crates/uv-distribution/src/distribution_database.rs:494]: I feel like I'm so allergic to unwrap that I'd probably do something like:

```rust
let progress = self
    .reporter
    .as_ref()
    .map(|reporter| (reporter, reporter.on_download_start(dist.name(), size)));
```

And then use `match self.progress` everywhere. What's your instinct?

File: crates/uv-distribution/src/distribution_database.rs [hunk 2]
```diff
@@ -813,6 +898,51 @@ impl<'a> ManagedClient<'a> {
     }
 }
 
+/// Returns the value of the Content-Length header for the request.
+fn content_length(req: &reqwest::Request) -> Option<u64> {
+    req.headers()
+        .get(reqwest::header::CONTENT_LENGTH)
+        .and_then(|val| val.to_str().ok())
+        .and_then(|val| val.parse::<usize>().ok())
```
Comment (charliermarsh) [inline crates/uv-distribution/src/distribution_database.rs:906]: Why not parse as `u64`? Just wondering.

**Comment by ibraheemdev** [comment]
@charliermarsh your design looks quite nice, I would be happy to switch to that.

**Comment by charliermarsh** [comment]
We can add a `--no-progress` flag?

**Comment by itayporezky** [comment]
After seeing this in the release notes i was afraid it would cause [this](https://stackoverflow.com/q/66290522) 
But in my testing it didn't, so that's great :)

**Pull request merged**